File: qt/gen_qdialog.go
package qt

import (
	"unsafe"
)

type QDialog__DialogCode int

const (
	QDialog__Rejected QDialog__DialogCode = 0
	QDialog__Accepted QDialog__DialogCode = 1
)

type QDialog struct {
	h          uintptr
	isSubclass bool
}

// NewQDialog constructs a new QDialog object.
func NewQDialog(parent *QWidget) *QDialog {
	ret := newQDialog(QDialog_new(parent.cPointer()))
	ret.isSubclass = true
	return ret
}

// NewQDialog2 constructs a new QDialog object.
func NewQDialog2() *QDialog {
	ret := newQDialog(QDialog_new2())
	ret.isSubclass = true
	return ret
}

// NewQDialog3 constructs a new QDialog object.
func NewQDialog3(parent *QWidget, f WindowType) *QDialog {
	ret := newQDialog(QDialog_new3(parent.cPointer(), (int)(f)))
	ret.isSubclass = true
	return ret
}

func (this *QDialog) MetaObject() *QMetaObject {
	return newQMetaObject(QDialog_MetaObject(this.h))
}

func (this *QDialog) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QDialog_Metacast(this.h, param1_Cstring))
}

func QDialog_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QDialog_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QDialog) Result() int {
	return (int)(QDialog_Result(this.h))
}

func (this *QDialog) SetVisible(visible bool) {
	QDialog_SetVisible(this.h, (bool)(visible))
}

func (this *QDialog) SizeHint() *QSize {
	_goptr := newQSize(QDialog_SizeHint(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QDialog) MinimumSizeHint() *QSize {
	_goptr := newQSize(QDialog_MinimumSizeHint(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QDialog) SetSizeGripEnabled(sizeGripEnabled bool) {
	QDialog_SetSizeGripEnabled(this.h, (bool)(sizeGripEnabled))
}

func (this *QDialog) IsSizeGripEnabled() bool {
	return (bool)(QDialog_IsSizeGripEnabled(this.h))
}

func (this *QDialog) SetModal(modal bool) {
	QDialog_SetModal(this.h, (bool)(modal))
}

func (this *QDialog) SetResult(r int) {
	QDialog_SetResult(this.h, (int)(r))
}

func (this *QDialog) Finished(result int) {
	QDialog_Finished(this.h, (int)(result))
}

func (this *QDialog) OnFinished(slot func(result int)) {
	QDialog_connect_Finished(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QDialog_Finished
func miqt_exec_callback_QDialog_Finished(cb intptr_t, result int) {
	gofunc, ok := cgo.Handle(cb).Value().(func(result int))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := (int)(result)

	gofunc(slotval1)
}

func (this *QDialog) Accepted() {
	QDialog_Accepted(this.h)
}

func (this *QDialog) OnAccepted(slot func()) {
	QDialog_connect_Accepted(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QDialog_Accepted
func miqt_exec_callback_QDialog_Accepted(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func (this *QDialog) Rejected() {
	QDialog_Rejected(this.h)
}

func (this *QDialog) OnRejected(slot func()) {
	QDialog_connect_Rejected(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QDialog_Rejected
func miqt_exec_callback_QDialog_Rejected(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func (this *QDialog) Open() {
	QDialog_Open(this.h)
}

func (this *QDialog) Exec() int {
	return (int)(QDialog_Exec(this.h))
}

func (this *QDialog) Done(param1 int) {
	QDialog_Done(this.h, (int)(param1))
}

func (this *QDialog) Accept() {
	QDialog_Accept(this.h)
}

func (this *QDialog) Reject() {
	QDialog_Reject(this.h)
}

func QDialog_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QDialog_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QDialog_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QDialog_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QDialog) callVirtualBase_MetaObject() *QMetaObject {
	return newQMetaObject(QDialog_virtualbase_MetaObject(unsafe.Pointer(this.h)))
}

func (this *QDialog) OnMetaObject(slot func(super func() *QMetaObject) *QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QDialog_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QDialog_MetaObject
func miqt_exec_callback_QDialog_MetaObject(self QDialog, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *QMetaObject) *QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QDialog{h: self}).callVirtualBase_MetaObject)

	return virtualReturn.cPointer()
}

func (this *QDialog) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QDialog_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QDialog) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QDialog_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QDialog_Metacast
func miqt_exec_callback_QDialog_Metacast(self QDialog, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)

	virtualReturn := gofunc((&QDialog{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}

File: qt/gen_qdialog.cpp
// +build ignore

#include <QCloseEvent>
#include <QContextMenuEvent>
#include <QDialog>
#include <QEvent>
#include <QKeyEvent>
#include <QMetaObject>
#include <QObject>
#include <QPaintDevice>
#include <QResizeEvent>
#include <QShowEvent>
#include <QSize>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QWidget>
#include <qdialog.h>
#include "gen_qdialog.h"
class MiqtVirtualQDialog : public virtual QDialog {
public:
MiqtVirtualQDialog(QWidget* parent): QDialog(parent) {};
MiqtVirtualQDialog(): QDialog() {};
MiqtVirtualQDialog(QWidget* parent, Qt::WindowFlags f): QDialog(parent, f) {};

virtual ~MiqtVirtualQDialog() = default;
// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;
// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
if (handle__MetaObject == 0) {
return QDialog::metaObject();
}
QMetaObject* callback_return_value = miqt_exec_callback_QDialog_MetaObject(const_cast<MiqtVirtualQDialog*>(this), handle__MetaObject);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QDialog::metaObject();

	}
// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;
// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
if (handle__Metacast == 0) {
return QDialog::qt_metacast(param1);
}
const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QDialog_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QDialog::qt_metacast(param1);

	}
};
QDialog* QDialog_new(QWidget* parent) {
return new MiqtVirtualQDialog(parent);
}
QDialog* QDialog_new2() {
return new MiqtVirtualQDialog();
}
QDialog* QDialog_new3(QWidget* parent, int f) {
return new MiqtVirtualQDialog(parent, static_cast<Qt::WindowFlags>(f));
}
void QDialog_virtbase(QDialog* src
, QWidget** outptr_QWidget
) {
*outptr_QWidget = static_cast<QWidget*>(src);
}
QMetaObject* QDialog_MetaObject(const QDialog* self) {
	return (QMetaObject*) self->metaObject();
}
void* QDialog_Metacast(QDialog* self, const char* param1) {
	return self->qt_metacast(param1);
}
struct miqt_string QDialog_Tr(const char* s) {
	QString _ret = QDialog::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
int QDialog_Result(const QDialog* self) {
	return self->result();
}
void QDialog_SetVisible(QDialog* self, bool visible) {
	self->setVisible(visible);
}
QSize* QDialog_SizeHint(const QDialog* self) {
	return new QSize(self->sizeHint());
}
QSize* QDialog_MinimumSizeHint(const QDialog* self) {
	return new QSize(self->minimumSizeHint());
}
void QDialog_SetSizeGripEnabled(QDialog* self, bool sizeGripEnabled) {
	self->setSizeGripEnabled(sizeGripEnabled);
}
bool QDialog_IsSizeGripEnabled(const QDialog* self) {
	return self->isSizeGripEnabled();
}
void QDialog_SetModal(QDialog* self, bool modal) {
	self->setModal(modal);
}
void QDialog_SetResult(QDialog* self, int r) {
	self->setResult(static_cast<int>(r));
}
void QDialog_Finished(QDialog* self, int result) {
	self->finished(static_cast<int>(result));
}
void QDialog_connect_Finished(QDialog* self, intptr_t slot) {
	MiqtVirtualQDialog::connect(self, static_cast<void (QDialog::*)(int)>(&QDialog::finished), self, [=](int result) {
		int sigval1 = result;
		miqt_exec_callback_QDialog_Finished(slot, sigval1);
	});
}
void QDialog_Accepted(QDialog* self) {
	self->accepted();
}
void QDialog_connect_Accepted(QDialog* self, intptr_t slot) {
	MiqtVirtualQDialog::connect(self, static_cast<void (QDialog::*)()>(&QDialog::accepted), self, [=]() {
		miqt_exec_callback_QDialog_Accepted(slot);
	});
}
void QDialog_Rejected(QDialog* self) {
	self->rejected();
}
void QDialog_connect_Rejected(QDialog* self, intptr_t slot) {
	MiqtVirtualQDialog::connect(self, static_cast<void (QDialog::*)()>(&QDialog::rejected), self, [=]() {
		miqt_exec_callback_QDialog_Rejected(slot);
	});
}
void QDialog_Open(QDialog* self) {
	self->open();
}
int QDialog_Exec(QDialog* self) {
	return self->exec();
}
void QDialog_Done(QDialog* self, int param1) {
	self->done(static_cast<int>(param1));
}
void QDialog_Accept(QDialog* self) {
	self->accept();
}
void QDialog_Reject(QDialog* self) {
	self->reject();
}
struct miqt_string QDialog_Tr2(const char* s, const char* c) {
	QString _ret = QDialog::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QDialog_Tr3(const char* s, const char* c, int n) {
	QString _ret = QDialog::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QDialog_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQDialog*>( (QDialog*)(self) )->handle__MetaObject = slot;
}
QMetaObject* QDialog_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQDialog*)(self) )->virtualbase_MetaObject();
}
void QDialog_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQDialog*>( (QDialog*)(self) )->handle__Metacast = slot;
}
void* QDialog_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQDialog*)(self) )->virtualbase_Metacast(param1);
}
void QDialog_Delete(QDialog* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQDialog*>( self );
	} else {
		delete self;
	}
}
